internal/certs/certs.go

```diff
@@ -0,0 +1,153 @@
+	err := r.generateSignerCerts()
+	if err != nil {
+		return err
+	}
+	adminCA := CertInfo{
+		commonName:      "admin-kubeconfig-signer",
+		certificatePath: filepath.Join(r.CertificatesDir, "admin-kubeconfig-client-ca.crt"),
+		keyPath:         filepath.Join(r.CertificatesDir, "admin-kubeconfig-client-ca.key"),
+	}
+	ca, err := generateAndWriteCA(adminCA)
+	if err != nil {
+		return err
+	}
+
+	r.userCert, r.userKey, err = generateAdminUserCertificate(ca)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Create signer keys and certificates and return a CA bundle from all CA certificates
+func (r *CertManager) generateSignerCerts() error {
+	var certAuthData []byte
+	for _, ci := range []CertInfo{
+		{
+			keyPath:    filepath.Join(r.CertificatesDir, "loadbalancer-serving-signer.key"),
+			commonName: "kube-apiserver-lb-signer",
+		},
+		{
+			keyPath:    filepath.Join(r.CertificatesDir, "localhost-serving-signer.key"),
+			commonName: "kube-apiserver-localhost-signer",
+		},
+		{
+			keyPath:    filepath.Join(r.CertificatesDir, "service-network-serving-signer.key"),
+			commonName: "kube-apiserver-service-network-signer",
+		},
+		{
+			keyPath:    filepath.Join(r.CertificatesDir, "ingresskey-ingress-operator.key"),
+			commonName: fmt.Sprintf("%s@%d", "ingress-operator", time.Now().Unix()),
+		},
+	} {
+		ca, err := generateAndWriteCA(ci)
+		if err != nil {
+			return err
+		}
+		certBytes, err := crypto.EncodeCertificates(ca.Config.Certs...)
+		if err != nil {
+			return err
+		}
+		// Append the PEM-encoded certificate to the bundle
+		certAuthData = append(certAuthData, certBytes...)
+	}
+	return nil
+}
+
+func (r *CertManager) GenerateKubeConfig(url string) ([]byte, error) {
+	kubeCfg := clientcmdapi.Config{
+		Kind:       "Config",
+		APIVersion: "v1",
+	}
+	kubeCfg.Clusters = map[string]*clientcmdapi.Cluster{
+		"cluster": {
+			Server:                   fmt.Sprintf("https://api.%s:6443", url),
+			CertificateAuthorityData: r.caBundle,
+		},
+	}
+	kubeCfg.AuthInfos = map[string]*clientcmdapi.AuthInfo{
+		"admin": {
+			ClientCertificateData: r.userCert,
+			ClientKeyData:         r.userKey,
+		},
+	}
+	kubeCfg.Contexts = map[string]*clientcmdapi.Context{
+		"admin": {
+			Cluster:   "cluster",
+			AuthInfo:  "admin",
+			Namespace: "default",
+		},
+	}
+	kubeCfg.CurrentContext = "admin"
+	return clientcmd.Write(kubeCfg)
+}
+
+func generateSelfSignedCACertificate(commonName string) (*crypto.CA, error) {
+	subject := pkix.Name{CommonName: commonName, OrganizationalUnit: []string{"openshift"}}
+	newCAConfig, err := crypto.MakeSelfSignedCAConfigForSubject(
+		subject,
+		crypto.DefaultCACertificateLifetimeInDays,
+	)
+	if err != nil {
+		return nil, errors.Wrap(err, "error generating self signed CA")
+	}
+	return &crypto.CA{
+		SerialGenerator: &crypto.RandomSerialGenerator{},
+		Config:          newCAConfig,
+	}, nil
+}
+
+func generateAndWriteCA(certInfo CertInfo) (*crypto.CA, error) {
```

Comment:
You're using this function for both the client CA and the kapi/ingress serving CAs - does the openshift installer create those certificates in an identical manner? Or could there be differences between them that we missed?

Reply:
I will need to verify that.

Reply:
After checking the certs generated by the openshift installer I found the following differences:
1. All certs but the ingress get 10 Years validity (ingress is valid for 2 years).
2. All the certs have `x509.KeyUsageCertSign`
I'll update the code to generate matching certs.

Reply:
TBH seems to me that the client cert shouldn't have x509.ExtKeyUsageServerAuth.
There's no scenario in which this certificate holder might get used for a server.
The purpose of this cert is to be used by a client while making requests to the kube-apiserver for authentication.
So, instead of adding `x509.ExtKeyUsageServerAuth` here I'll send a PR to the installer for removing the serverAuth